test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ClientTests/Tests/ClientQueryTests.cs

```diff
@@ -0,0 +1,228 @@
+
+            //GET http://jinfutanodata01:9090/AstoriaDefault635157551526070289/Person()?$filter=PersonId%20le%20256 HTTP/1.1
+            //all the IDs in [1, 2] are counted in.
+            result = _context.People.Where(c => c.PersonId > 0);
+            Assert.Equal(2, result.ToList().Count());
+        }
+
+        [Fact]
+        public void InvalidDollarFilter_UsingContains_ThrowsExceptions()
+        {
+            string[] errorUrls =
+            {
+                "Logins?$filter=contains(Username, 1)",
+                "People?$filter=contains(Name, \"m\")",
+                "Cars?$filter=contains(VIN, '12')"
+            };
+
+            foreach (var errorUrl in errorUrls)
+            {
+                try
+                {
+                    _context.Execute<Common.Clients.EndToEnd.Person>(new Uri(_baseUri.OriginalString + errorUrl));
+                    Assert.True(false, "Expected Exception not thrown for " + errorUrl);
+                }
+                catch (DataServiceQueryException ex)
+                {
+                    Assert.Equal(400, ex.Response.StatusCode);
+                }
+            }
+        }
+
+        [Fact]
+        public void GetAllPagesTest()
+        {
+            int allCustomersCount = _context.Customers.Count();
+            bool CheckNextLink = false;
+            Uri nextPageLink = null;
+
+            EventHandler<SendingRequest2EventArgs> sendRequestEvent = (sender, args) =>
+            {
+                //The first request should not be checked.
+                if (CheckNextLink)
+                {
+                    Assert.Equal(nextPageLink.AbsoluteUri, args.RequestMessage.Url.AbsoluteUri);
+                }
+                CheckNextLink = true;
+            };
+
+            _context.Configurations.ResponsePipeline.OnFeedEnded((args) =>
+            {
+                nextPageLink = args.Feed.NextPageLink;
+            });
+
+            _context.SendingRequest2 += sendRequestEvent;
+            int queryCustomersCount = _context.Customers.GetAllPages().Count();
+            Assert.Equal(allCustomersCount, queryCustomersCount);
+
+            //$filter
+            _context.SendingRequest2 -= sendRequestEvent;
+            var filterCustomersCount = _context.Customers.Where(c => c.CustomerId > -5).Count();
+
+            _context.SendingRequest2 += sendRequestEvent;
+            CheckNextLink = false;
+            queryCustomersCount = ((DataServiceQuery<Common.Clients.EndToEnd.Customer>)_context.Customers.Where(c => c.CustomerId > -5)).GetAllPages().ToList().Count();
+            Assert.Equal(filterCustomersCount, queryCustomersCount);
+
+            //$projection
+            CheckNextLink = false;
+            queryCustomersCount = ((DataServiceQuery<Common.Clients.EndToEnd.Customer>)_context.Customers.Select(c => new Common.Clients.EndToEnd.Customer() { CustomerId = c.CustomerId, Name = c.Name })).GetAllPages().ToList().Count();

```

Comment:
What's the purpose of the `ToList()` call? Why not:
```suggestion
            queryCustomersCount = ((DataServiceQuery<Common.Clients.EndToEnd.Customer>)_context.Customers.Select(c => new Common.Clients.EndToEnd.Customer() { CustomerId = c.CustomerId, Name = c.Name })).GetAllPages().Count();

```